server/src/main/java/org/apache/druid/server/QueryLifecycle.java

```diff
@@ -82,13 +91,43 @@
   private final RequestLogger requestLogger;
   private final AuthorizerMapper authorizerMapper;
   private final DefaultQueryConfig defaultQueryConfig;
+  private final AuthConfig authConfig;
   private final long startMs;
   private final long startNs;
 
   private State state = State.NEW;
   private AuthenticationResult authenticationResult;
   private QueryToolChest toolChest;
-  private Query baseQuery;
+
+  /**
+   * A holder for the user query to run.
+   *
+   * The holder has a state for query context.
+   * The query context in {@link QueryHolder#delegate#context} is not valid until they are authorized.
+   * {@link #context} should be used instead to get query context until authorized.
+   *
+   * Variable state change flow:
+   *
+   * - Initialized to null.
+   * - Set in {@link #initialize}.
+   * - Updated with context after authorization in {@link #doAuthorize}.
+   */
+  @MonotonicNonNull
+  private QueryHolder<?> baseQuery;
```

Comment:
this seems error prone to have two contexts like this with one hidden in the wrapper, I think it might be easier to follow what is going on here to keep this as a `Query` and have a `baseContext` separate from the `context`.